src/dsl.jl

```diff
@@ -516,155 +450,208 @@ end
+# Takes a ModelingToolkit declaration macro (like @parameters ...) and return and expression:
+# That calls the macro and then scalarizes all the symbols created into a vector of Nums.
+# stores the created symbolic variables in a variable (which name is generated from `name`).
+# It will also return the name used for the variable that stores the symbolic variables.
+function scalarize_macro(expr_init, name)
+    # Generates a random variable name which (in generated code) will store the produced
+    # symbolic variables (e.g. `var"##ps#384"`).
     namesym = gensym(name)
-    if nonempty
-        if scalarize
-            symvec = gensym()
-            ex = quote
-                $symvec = $ex
-                $namesym = reduce(vcat, Symbolics.scalarize($symvec))
-            end
-        else
-            ex = quote
-                $namesym = $ex
-            end
+
+    # If the input expression is non-empty, wraps it with additional information.
+    if expr_init != :(())
+        symvec = gensym()
+        expr = quote
+            $symvec = $expr_init
+            $namesym = reduce(vcat, Symbolics.scalarize($symvec))
         end
     else
-        ex = :($namesym = Num[])
+        expr = :($namesym = Num[])
     end
-    ex, namesym
+
+    return expr, namesym
+end
+
+# From the system reactions (as `DSLReaction`s) and equations (as expressions),
+# creates the expressions that evaluates to the reaction (+ equations) vector.
+function get_rxexprs(reactions, equations, diffsyms)
+    rxsexprs = :(Catalyst.CatalystEqType[])
+    foreach(rx -> push!(rxsexprs.args, get_rxexpr(rx)), reactions)
+    foreach(eq -> push!(rxsexprs.args, escape_equation!(eq, diffsyms)), equations)
+    return rxsexprs
+end
+
+# From a `DSLReaction` struct, creates the expression which evaluates to the creation
+# of the corresponding reaction.
+function get_rxexpr(rx::DSLReaction)
+    # Initiates the `Reaction` expression.
+    rate = recursive_escape_functions!(rx.rate)
+    rx_constructor = :(Reaction($rate, nothing, nothing; metadata = $(rx.metadata)))
+
+    # Loops through all products and substrates, and adds them (and their stoichiometries)
+    # to the `Reaction` expression.
+    for sub in rx.substrates
+        push!(rx_constructor.args[4].args, sub.reactant)
+        push!(rx_constructor.args[6].args, sub.stoichiometry)
+    end
+    for prod in rx.products
+        push!(rx_constructor.args[5].args, prod.reactant)
+        push!(rx_constructor.args[7].args, prod.stoichiometry)
+    end
+
+    return rx_constructor
+end
+
+# Recursively escape functions within equations of an equation written using user-defined functions.
+# Does not expand special function calls like "hill(...)" and differential operators.
+function escape_equation!(eqexpr::Expr, diffsyms)
+    eqexpr.args[2] = recursive_escape_functions!(eqexpr.args[2], diffsyms)
+    eqexpr.args[3] = recursive_escape_functions!(eqexpr.args[3], diffsyms)
+    eqexpr
 end
```

Comment:
Moved `escape_equation!` up here since it is just a helper function to the function above and not involved in other stuff. Calls `recursive_escape_functions!` on the lhs as well.